index.js

```diff
@@ -1,123 +1,133 @@
+const createDatabaseConnection = dbName => {
     debug('creating connection for db "%s"', dbName)
 
-    var configCopy = Object.assign({}, serverConfig)
-    configCopy.database = dbName
-    createConnection(configCopy, (error, connection) => {
-        if (error) {
-            return callback(error)
-        }
-
-        connectionCache[dbName] = connection
-        callback(null, connection)
-    })
+    connectionCache[dbName] = mongoClient.db(dbName)
+    return connectionCache[dbName]
 }
 
 const getConnection = (dbName, callback) => {
     if (typeof dbName === 'function') {
         callback = dbName
         dbName = 'testDatabase'
     }
-
-    var connection = connectionCache[dbName]
-    if (connection) {
-        debug('retrieve connection from connection cache for db "%s"', dbName)
-        return process.nextTick(() => callback(null, connection))
+    if (!dbName){
+      dbName = 'testDatabase'
     }
 
-    if (serverConfig) {
-        return createServerSpecificConfiguration(serverConfig, dbName, callback)
+    if (initPromise === null){
+        initPromise = startServer().then(createMongoClient)
     }
 
-    startServer((error, resultConfiguration) => {
-        if (error) {
-            return callback(error)
-        }
-
-        serverConfig = resultConfiguration
-        createServerSpecificConfiguration(serverConfig, dbName, callback)
-    })
+    return initPromise
+        .then(() => createDatabaseConnection(dbName))
+        .then(connection => {
+            if (typeof callback === 'function'){
+                callback(null, connection)
+            } else {
+                return connection
+            }
+        }).catch(err => {
+            if (typeof callback === 'function'){
+                callback(err)
+            } else {
+                throw err
+            }
+        })
 }
 
 const shutDown = callback => {
-    if (typeof callback !== 'function') {
-        callback = () => {}
-    }
-
-    if (serverEmitter) {
-        debug('emit shutdown event')
-        serverEmitter.emit('mongoShutdown')
-    }
-
-    serverEmitter = null
-    serverConfig = null
-
-    var connections = Object.keys(connectionCache).map(key => connectionCache[key])
-    if (connections.length > 0) {
-        debug('closing %d mongo connections', connections.length)
-        async.each(connections, (con, cb) => con.close(cb), error => {
-            connectionCache = {}
-            callback(error)
+    let cleanup = Promise.resolve()
+
+    if (initPromise !== null) {
+        cleanup = initPromise.then(() => {
+            debug('closing mongo client')
+            return mongoClient.close(true)
+        }).then(() => {
+            mongoClient = null
+            debug('sending mongod process')
+            mongodHelper.kill()
+        }).then(() => {
+            // mongodb-prebuilt doesnt provide a mechanism itself to notify when shutdown really happened
```

Comment:
This block could be made optional, since it relies on the internals of mongodb-prebuilt. Maybe allowing a flag to be passed onto the cleanup method to avoid waiting for the close event